### ISSUE TITLE
[Bug] New "Hostnames plugin" does not seem to work at all

### ISSUE THREAD

**Issue by saloniamatteo** [searxng/searxng#3558]
TL;DR; [the "unhashable type" error is a result of mixing up YAML's lists and mappings.](https://github.com/searxng/searxng/issues/3558#issuecomment-2175058128) .. online docs has been improved .. read [online doc](https://docs.searxng.org/src/searx.plugins.hostnames.html) carefully  ..

------------

SearXNG version: 2024.6.7+f5eb56b63

**What happened?**
After the switch from "hostname_replace" to the new "Hostnames plugin", it doesn't seem to work anymore.

Modifying settings.yml, and searching for a relevant query, shows us that the plugin does not seem to be working. (Replacing does not work either, nor do priorities.)

**How To Reproduce**
Edit `settings.yml`, and add the following:

```yaml
enabled_plugins:
 - 'Hostnames plugin'
```

```yaml
hostnames:
 remove:
 - '(.*\.)?example.com$'
```

**Expected behavior**
The results for `example.com` and its subdomains should not appear at all.

**Screenshots & Logs**
![Screenshot_20240609_183858_Termux](https://github.com/searxng/searxng/assets/28765699/dc80b564-48bb-445a-8dfc-4837d4f1a0ae)

(Just to be sure, I added 'Hostname replace' as well, however, it didn't make a difference)

![Screenshot_20240609_183947_Mull](https://github.com/searxng/searxng/assets/28765699/90ae8ff6-6098-49b3-87f0-8b09a510d719)

**Comment by saloniamatteo** [comment]
Upon further inspection, I got the following error:

![Screenshot_20240609_190351_Termux](https://github.com/searxng/searxng/assets/28765699/74ad1289-4261-42a2-b582-2db592a4bb88)

I do not know why the plugin appears to be disabled, even though I added both names (old & new).

Edit: I want to make it clear this error happens even when removing the old deprecated name.

**Comment by return42** [comment]
Its disabled because the setting has an issue .. check indentations:

```yaml
hostnames:
 remove:
 - '(.*\.)?example.com$'
```
fixed by indentaion of the list items .. (the last line):

```yaml
hostnames:
 remove:
   - '(.*\.)?example.com$'
```

**Comment by saloniamatteo** [comment]
@return42 I'm afraid this did not fix the issue.

**Comment by return42** [comment]
I'm pretty sure you have problems with the indentations or something like that ...

**Comment by saloniamatteo** [comment]
I can confirm I do not have indentation problems, it would be idiotic to have a whitespace-sensitive config file... The issue persists.

**Comment by return42** [comment]
Please give us your config .. and the traceback .. but please don't send screenshots.

> The issue persists.

.. have a doubt, since I have configured hostname plugin on my instance without the issue you reported here.

**Comment by saloniamatteo** [comment]
Here are the exact lines I put in settings.yml:

```yaml
hostnames:
   remove:
        - '(.*\.)?example.com'
        - '(.*\.)?example.com$'
        - 'example.com$'
```

None of these variations work.

**Comment by saloniamatteo** [comment]
Trace:

```
2024-06-10 14:57:20,166 ERROR:searx.plugins: searx.plugins.hostnames: exception while loading, the plugin is disabled
Traceback (most recent call last):
  File "/usr/local/searxng/searxng-src/searx/plugins/__init__.py", line 115, in load_plugin
    plugin = import_module(plugin_module_name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/local/searxng/searxng-src/searx/plugins/hostnames.py", line 120, in <module>
    replacements = _load_regular_expressions_with_fallback('replace')
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searxng-src/searx/plugins/hostnames.py", line 102, in _load_regular_expressions_with_fallback
    expressions = _load_regular_expressions(settings_key)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searxng-src/searx/plugins/hostnames.py", line 91, in _load_regular_expressions
    return {re.compile(r) for r in setting_value}
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searxng-src/searx/plugins/hostnames.py", line 91, in <setcomp>
    return {re.compile(r) for r in setting_value}
            ^^^^^^^^^^^^^
  File "/usr/lib/python3.11/re/__init__.py", line 227, in compile
    return _compile(pattern, flags)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/re/__init__.py", line 277, in _compile
    return _cache[type(pattern), pattern, flags]
           ~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unhashable type: 'dict'
```

**Comment by return42** [comment]
> Here are the exact lines I put in settings.yml:
> 
> ```yaml
> hostnames:
>    remove:
>         - '(.*\.)?example.com'
>         - '(.*\.)?example.com$'
>         - 'example.com$'
> ```
> 
> None of these variations work.

Worked for me .. I assume this is not your complete hostname setup?

> it would be idiotic to have a whitespace-sensitive config file

its yaml .. https://en.wikipedia.org/wiki/YAML

![image](https://github.com/searxng/searxng/assets/554536/2323fd3f-46cc-44bd-99dd-53e998219a05)

**Comment by saloniamatteo** [comment]
I still don't understand why it does not work. I have updated & rebuilt the python venv, thinking that may have been the culprit, but nothing's changed.

> I assume this is not your complete hostname setup?

It is, for now, just for testing purposes. Of course, I enabled "Hostnames plugin" under "enabled_plugins".

**Comment by saloniamatteo** [comment]
Out of the blue, it started working!

@return42 thanks for your insight regarding YAML's retarded whitespace-sensitivity, I did not know a space could break an entire functionality. 😥

**Comment by omentic** [comment]
For posterity for anyone searching here, the "unhashable type" error is a result of mixing up YAML's lists and mappings. The `replace` directive expects a *mapping*, with entries not prefixed with `-`. The other directives all expect *lists*, with entries prefixed with `-`. A configuration should look like the following:

```yaml
hostnames:
 remove: 
   - '(.*\\.)?facebook.com$'
   ...
 low_priority:
   - '(.*\\.)?google(\\..*)?$'
   ...
 high_priority:
   - '(.*\\.)?wikipedia.org$'
   ...
 replace:
   '(.*\\.)?youtube\\.com$': 'invidious.example.com'
   '(.*\\.)?youtu\\.be$': 'invidious.example.com'
   ...

```

I spent longer than I'd care to admit tracking this down...

**Comment by return42** [comment]
> For posterity for anyone searching here, the "unhashable type" error is a result of mixing up YAML's lists and mappings.

I improved the online docs:

- https://github.com/searxng/searxng/pull/3585

**Comment by mbirth** [comment]
> Out of the blue, it started working!

Just to add for anyone with similar issues: For me, the *Hostnames plugin* showed up as enabled in the Preferences view, however, it seems to have been in an undefined state. Only after toggling the switch to off and then on again and saving the preferences, my filters started working.

Also, I am using a [redirect](https://mshibanami.github.io/redirect-web/) (to use SearXNG from Safari) that included my preferences hash - so that redirect had to be updated as well after toggling the switch.
